File: python/tidegraph/region.py
"""Functional selectors with complete candidate controls and region-owned history."""
from dataclasses import dataclass
from types import MappingProxyType
import torch
from .history import History, increment, validate as validate_history


@dataclass(frozen=True)
class RegionLayout:
    spec: object
    members: tuple[int, ...]
    slots: object

    @classmethod
    def create(cls, spec, members):
        members = tuple(members)
        return cls(spec, members, MappingProxyType({v: i for i, v in enumerate(members)}))


@dataclass(frozen=True)
class RegionInput:
    history: History
    time: int
    candidates: tuple[tuple[int, torch.Tensor], ...]
    layout: RegionLayout
    payload_dtype: torch.dtype
    payload_device: torch.device


@dataclass
class Selection:
    active: set[int]
    controls: dict[int, torch.Tensor]
    history: History


class RegionProgram(torch.nn.Module):
    profile = "custom"

    def initial(self, layout, reference):
        return History()

    def step(self, request):
        raise NotImplementedError

    def validate_history(self, history, layout):
        pass

    def validate_weights(self, layout):
        pass


class CountSelector(RegionProgram):
    profile = "count-v1"

    def initial(self, layout, reference):
        return History(node_maps={"selected": {}})

    def scores(self, request):
        return [d for _, d in request.candidates]

    def eligible(self, score):
        return True

    def step(self, r):
        nodes = [v for v, _ in r.candidates]
        scores = self.scores(r)
        if any(not torch.isfinite(d).all() for d in scores):
            raise ValueError("nonfinite region selector score")
        counts = r.history.node_maps["selected"]
        order = sorted((i for i, d in enumerate(scores) if self.eligible(d)), key=lambda i: (
            counts.get(nodes[i], 0) if r.layout.spec.count_priority else 0, -float(scores[i].detach()), nodes[i]))
        active = {nodes[i] for i in order[:r.layout.spec.budget]}
        probs = torch.stack(scores).softmax(0).to(dtype=r.payload_dtype, device=r.payload_device)
        history = r.history.fork()
        history.last_time = r.time
        for v in active:
            history.node_maps["selected"][v] = increment(counts.get(v, 0))
        return Selection(active, dict(zip(nodes, probs.unbind(0))), history)

    def validate_history(self, history, layout):
        if (history.scalars or history.node_maps.keys() != {"selected"}
                or any(c < 0 for c in history.node_maps["selected"].values())):
            raise ValueError("invalid selected-count history layout")
        if history.tensors:
            raise ValueError("unexpected count-selector history tensors")


class PositiveSelector(CountSelector):
    """Simple empty-selection anchor: only strictly positive scores are eligible."""
    profile = "positive-v1"

    def eligible(self, score):
        return float(score.detach()) > 0


class TensorHistorySelector(CountSelector):
    profile = "tensor-history-v1"

    def __init__(self, members, dtype):
        super().__init__()
        self.alpha = torch.nn.Parameter(torch.tensor(.5, dtype=dtype))
        self.bias = torch.nn.Parameter(torch.linspace(-.2, .2, members, dtype=dtype))

    def initial(self, layout, reference):
        return History(node_maps={"selected": {}}, tensors={"memory": reference.new_zeros(())})

    def scores(self, r):
        memory = r.history.tensors["memory"]
        return [d + memory*self.bias[r.layout.slots[v]] for v, d in r.candidates]

    def step(self, r):
        result = super().step(r)
        result.history.tensors["memory"] = self.alpha*r.history.tensors["memory"] + torch.stack(
            [d for _, d in r.candidates]).sum().to(dtype=r.payload_dtype, device=r.payload_device)
        return result

    def validate_weights(self, layout):
        if self.alpha.shape != () or self.bias.shape != (len(layout.members),):
            raise ValueError("region parameter membership layout mismatch")

    def validate_history(self, history, layout):
        if (history.scalars or history.node_maps.keys() != {"selected"}
                or any(c < 0 for c in history.node_maps["selected"].values())
                or history.tensors.keys() != {"memory"} or history.tensors["memory"].shape != ()):
            raise ValueError("invalid tensor-selector history layout")


def program(layout, dtype):
    profile = layout.spec.selector
    if profile == CountSelector.profile:
        return CountSelector()
    if profile == PositiveSelector.profile:
        return PositiveSelector()
    if profile == TensorHistorySelector.profile:
        return TensorHistorySelector(len(layout.members), dtype)
    from .lh_selector import LHSelector
    if profile == LHSelector.profile:
        return LHSelector()
    raise ValueError("unknown region selector profile")


def validate_program(program, layout, reference, *, native=False):
    from .lh_selector import LHSelector
    if native and type(program) not in {CountSelector, PositiveSelector, TensorHistorySelector, LHSelector}:
        raise ValueError("Python custom region selector has no native implementation")
    if not isinstance(program, RegionProgram) or program.profile != layout.spec.selector:
        raise ValueError("region program does not match graph profile")
    for value in (*program.parameters(), *program.buffers()):
        if (value.dtype, value.device) != (reference.dtype, reference.device) or not torch.isfinite(value).all():
            raise ValueError("incompatible region parameter dtype/device/value")
    program.validate_weights(layout)


def evaluate(graph, model, q, batch, region, time, candidates):
    """Only called for a complete, canonical nonempty candidate set."""
    layout, p = graph.region_layouts[region], model.regions[region]
    reference = model.nodes[0].bias
    nodes = [v for v, _ in candidates]
    if not nodes or nodes != sorted(set(nodes)) or any(v not in layout.slots for v in nodes):
        raise ValueError("invalid region candidate domain/order")
    old = q.history.get((batch, region))
    if old is None:
        old = p.initial(layout, reference)
        validate_history(old, layout, reference, time-1)
        p.validate_history(old, layout)
    result = p.step(RegionInput(old, time, tuple(candidates), layout, reference.dtype, reference.device))
    if (not isinstance(result, Selection) or not isinstance(result.active, set)
            or any(type(v) is not int for v in result.active) or not result.active <= set(nodes)
            or len(result.active) > layout.spec.budget):
        raise ValueError("region selector returned invalid active subset/capacity")
    if (not isinstance(result.controls, dict) or result.controls.keys() != set(nodes)
            or any(type(v) is not int for v in result.controls)):
        raise ValueError("region selector returned invalid control domain")
    for control in result.controls.values():
        if not isinstance(control, torch.Tensor) or (control.dtype, control.device) != (reference.dtype, reference.device):
            raise ValueError("region selector returned incompatible control dtype/device")
        if not torch.isfinite(control).all():
            raise ValueError("region selector returned nonfinite control")
    validate_history(result.history, layout, reference, time)
    p.validate_history(result.history, layout)
    q.history[batch, region] = result.history
    return result.active, result.controls, result.history

File: cpp/src/region.cpp
#include "tide/region.h"
#include "tide/counters.h"
#include <algorithm>
#include <cmath>
#include <stdexcept>

namespace tide {
Index RegionLayout::slot(Index node) const {
  auto it = std::lower_bound(members.begin(), members.end(), node);
  if (it == members.end() || *it != node) throw std::invalid_argument("node is outside region membership");
  return it - members.begin();
}
RegionLayout region_layout(const Graph& g, Index region) {
  auto begin = g.region_index.offsets[region], end = g.region_index.offsets[region+1];
  return {g.regions[region], c10::ArrayRef<Index>(g.region_index.edges).slice(begin, end-begin)};
}
namespace {
class CountSelector final : public RegionKernel {
 public:
  explicit CountSelector(std::string profile) : profile_(std::move(profile)) {}
  History initial(const RegionWeights&, const RegionLayout&, const Tensor& ref) const override {
    History h; h.node_maps["selected"] = {};
    if (tensor_history()) h.tensors["memory"] = at::zeros({}, ref.options());
    return h;
  }
  Selection step(const RegionWeights& w, const RegionInput& r) const override {
    const auto& counts = r.history.node_maps.at("selected");
    std::vector<Tensor> scores;
    std::vector<std::tuple<Index, double, Index>> ranking;
    for (const auto& candidate : r.candidates) {
      const auto node = candidate.node;
      auto score = candidate.descriptor;
      if (tensor_history()) score = score + r.history.tensors.at("memory")*w.extra.at("bias")[r.layout.slot(node)];
      auto value = score.item<double>();
      if (!std::isfinite(value)) throw std::invalid_argument("nonfinite region selector score");
      auto count = counts.find(node);
      if (profile_ != "positive-v1" || value > 0)
        ranking.emplace_back(r.layout.spec.count_priority && count != counts.end() ? count->second : 0, -value, node);
      scores.push_back(score);
    }
    std::sort(ranking.begin(), ranking.end());
    Selection result; result.history = r.history; result.history.last_time = r.time;
    for (Index i = 0; i < std::min<Index>(r.layout.spec.budget, ranking.size()); ++i) {
      auto node = std::get<2>(ranking[i]); result.active.insert(node);
      auto& count = result.history.node_maps.at("selected")[node]; count = increment(count);
    }
    auto controls = at::softmax(at::stack(scores), 0).to(r.payload_options);
    for (size_t i = 0; i < scores.size(); ++i) result.controls.emplace(r.candidates[i].node, controls[i]);
    if (tensor_history()) {
      std::vector<Tensor> descriptors;
      for (const auto& candidate : r.candidates) descriptors.push_back(candidate.descriptor);
      result.history.tensors["memory"] = w.extra.at("alpha")*r.history.tensors.at("memory") + at::stack(descriptors).sum().to(r.payload_options);
    }
    return result;
  }
  void validate_weights(const RegionWeights& w, const RegionLayout& layout) const override {
    if (!tensor_history()) {
      if (!w.extra.empty()) throw std::invalid_argument("unexpected count-selector parameters");
      return;
    }
    if (w.extra.size() != 2 || !w.extra.count("alpha") || !w.extra.count("bias")
        || w.extra.at("alpha").dim() != 0 || w.extra.at("bias").sizes() != at::IntArrayRef{static_cast<Index>(layout.members.size())})
      throw std::invalid_argument("region parameter membership layout mismatch");
  }
  void validate_history(const History& h, const RegionLayout&) const override {
    if (!h.scalars.empty() || h.node_maps.size() != 1 || !h.node_maps.count("selected"))
      throw std::invalid_argument("invalid selected-count history layout");
    for (const auto& [node, count] : h.node_maps.at("selected"))
      if (count < 0) throw std::invalid_argument("invalid selected-count history value");
    if (tensor_history()) {
      if (h.tensors.size() != 1 || !h.tensors.count("memory") || h.tensors.at("memory").dim() != 0)
        throw std::invalid_argument("invalid tensor-selector history layout");
    } else if (!h.tensors.empty()) throw std::invalid_argument("unexpected count-selector history tensors");
  }
 private:
  bool tensor_history() const { return profile_ == "tensor-history-v1"; }
  std::string profile_;
};
}  // namespace
std::shared_ptr<const RegionKernel> make_lh_selector();
std::shared_ptr<const RegionKernel> make_region_kernel(const Region& spec) {
  if (spec.selector == "lh-count-affect-v1") return make_lh_selector();
  if (spec.selector == "count-v1" || spec.selector == "positive-v1" || spec.selector == "tensor-history-v1")
    return std::make_shared<CountSelector>(spec.selector);
  throw std::invalid_argument("unknown region selector profile: " + spec.selector);
}
}  // namespace tide
